// pilt-lib-chat/modules/leaguedata/index.js
var events = require('events');

function TimtasticLolApi(mashapeKey, requestPackage) {
  var api = this;
  events.EventEmitter.call(api);
  var key = mashapeKey;
  var request = requestPackage;

  // Constants
  var apiUrl = 'https:\/\/community-league-of-legends.p.mashape.com';

  var call = function (apiPath, callback) {
    var options = {
      url: apiUrl + apiPath,
      json: true,
      method: 'GET',
      headers: { 'X-Mashape-Authorization': key }
    };
    request(options, function(error, response, body) {
      if (error) {
        api.emit('error', response.statusCode, apiPath, error);
      } else {
        if (response.statusCode !== 200) {
          api.emit('error', response.statusCode, apiPath);
        } else {
          api.emit('success', body, apiPath);
          callback(body);
        }
      }
    });
  };

  // Endpoints
  api.getAggregatedStats = function (region, accountId, callback) {
    return call('/api/v1.0/'+region+'/summoner/getAggregatedStats/'+accountId, callback);
  };
  api.getAllPublicSummonerDataByAccount = function (region, accountId, callback) {
    return call('/api/v1.0/'+region+'/summoner/getAllPublicSummonerDataByAccount/'+accountId, callback);
  };
  api.getLeagueForPlayer = function (region, summonerId, callback) {
    return call('/api/v1.0/'+region+'/summoner/getLeagueForPlayer/'+summonerId, callback);
  };
  api.getRecentGames = function (region, accountId, callback) {
    return call('/api/v1.0/'+region+'/summoner/getRecentGames/'+accountId, callback);
  };
  api.getSummonerByName = function (region, summonerName, callback) {
    return call('/api/v1.0/'+region+'/summoner/getSummonerByName/'+summonerName, callback);
  };
  api.getSummonerBySummonerId = function (region, summonerId, callback) {
    return call('/api/v1.0/'+region+'/summoner/getSummonerBySummonerId/'+summonerId, callback);
  };
  api.getSummonerHonor = function (region, summonerId, callback) {
    return call('/api/v1.0/'+region+'/summoner/getSummonerHonor/'+summonerId, callback);
  };
  api.retrieveInProgressSpectatorGameInfo = function (region, summonerName, callback) {
    return call('/api/v1.0/'+region+'/summoner/retrieveInProgressSpectatorGameInfo/'+summonerName, callback);
  };
  api.retrievePlayerStatsByAccountId = function (region, accountId, callback) {
    return call('/api/v1.0/'+region+'/summoner/retrievePlayerStatsByAccountId/'+accountId, callback);
  };
}
TimtasticLolApi.prototype.__proto__ = events.EventEmitter.prototype;
module.exports = TimtasticLolApi;
